Initial pass at annotations chapter

djnavarro opened the pull request:
I've started work on a rethink to the Annotations chapter. It still needs polish (see my comments below), but I think I've has reached the point where you may wish to comment... and suggest a different direction if this approach doesn't seem right to you! 

The structure of the chapter currently has five sections, but I am tentatively thinking I'll collapse it to four:

- **Adding text labels to points** contains most of the old section on text labels, and adds an example of nice labelling using ggrepel (the previous example with directlabel has moved)

- **Building custom annotations** is the old section that was just labelled Annotations, with a slight rewrite to emphasise that the goal here is merely to illustrate how to build your own annotations using the basic geoms, and ends by describing the annotate() helper function

- **Replacing a legend with text** moves the previous example based on directlabel into its own section and adds a second example using ggforce. My logic here was that the directlabel example felt a little different from the other text label ones -- it's not really labelling specific points its highlighting a group of points. I think it works better with this as a separate section, particularly when accompanied with the geom_mark_ellipse() example. (Possibly this section could be renamed... highlighting groups; and the gghighlight example could be moved here???)

- **Reference lines across facets** has the previous example using the diamonds data, just slightly expanded for clarity. At the moment this feels a little orphaned, and I'm wondering if it would be sensible to rename this "Annotation across facets" and maybe give another example? It might work particularly well if I move the first gghighlight example up to the preceding section, and then talk about faceting with gghighlight? 

- **Other annotation tools** I'm not entirely happy with having an "other" to be honest, so I'm tempted to move the gghighlight example and drop the reference to cowplot? I have a suspicion that it will turn out to be worth referencing cowplot in some other section.

The other thing I considered discussing here is patchwork, but I feel like that's something else? It doesn't seem to belong in a chapter on "Annotation" specifically? 

Let me know what you think of this general plan - I can continue tidying or go in a different direction :-)

hadley commented on annotations.Rmd:
```diff
@@ -5,15 +5,27 @@ source("common.R")
-Adding text to a plot can be quite tricky. ggplot2 doesn't have all the answers, but does provide some tools to make your life a little easier. The main tool is `geom_text()`, which adds `label`s at the specified `x` and `y` positions. 
 
-`geom_text()` has the most aesthetics of any geom, because there are so many ways to control the appearance of a text:
+## Adding Text Labels to Points {#sec:labelling}
```
Can you please use sentence case in headings?

And maybe just call "Text labels"?

hadley commented on annotations.Rmd:
```diff
@@ -92,10 +112,11 @@ You can map data values to these aesthetics, but use restraint: it is hard to pe
+*   The third parameter is `check_overlap`. If `check_overlap = TRUE`, overlapping 
+    labels will be automatically removed from the plot. The algorithm is simple: 
+    labels are plotted in the order they appear in the data frame; if a label 
+    would overlap with an existing point, it's omitted. This is not incredibly useful, 
```
@paleolimbot realised that this is actually much more useful than I realised because you can order the labels from most important to least important.

djnavarro commented:
Updated version incorporates all of your suggested changes (I think!). I've added an example of underplotting to emphasise a subset (the subaru example), and folded in an example of annotating with geom_curve in the process (inspired by the bbc cookbook).

hadley commented on annotations.Rmd:
```diff
@@ -67,19 +83,23 @@ Adding text to a plot can be quite tricky. ggplot2 doesn't have all the answers,
 *   `angle` specifies the rotation of the text in degrees.
 
-You can map data values to these aesthetics, but use restraint: it is hard to percieve the relationship between variables mapped to these aesthetics. `geom_text()` also has three parameters. Unlike the aesthetics, these only take single values, so they must be the same for all labels:
+The ggplot2 package does allow you to map data values to the aesthetics used but `geom_text()`, but you should use restraint: it is hard to perceive the relationship between variables mapped to these aesthetics, and rarely useful to do so. 
```
```suggestion
The ggplot2 package does allow you to map data values to the aesthetics used by `geom_text()`, but you should use restraint: it is hard to perceive the relationship between variables mapped to these aesthetics, and rarely useful to do so. 
```

djnavarro replied:
thanks!

hadley commented on annotations.Rmd:
```diff
@@ -133,33 +157,30 @@ Labelling data well poses some challenges:
+    positioning to avoid overlap. It works quite well so long as the number 
+    of labels is not excessive:
+    
+`r columns(1, 2/3)`
```
Should this be indent to align with the bullet above?

djnavarro replied:
Yep. Fixed.

hadley commented on annotations.Rmd:
````diff
@@ -228,28 +249,113 @@ ggplot(economics, aes(date, unemploy)) +
+
+`r columns(1, 2/3)`
+```{r underplot}
+p + annotate(geom = "point", x = 5.5, y = 40, 
````
I think these should fit in a single line? 

If they don't, can you please use the tidyverse long lines style? http://style.tidyverse.org/syntax.html#long-lines

djnavarro replied:
Fixed in the current iteration, along with several others I noticed

hadley replied:
Btw when you resolve a query, you can click the "resolve conversation" button and I'll trust you 😄

hadley commented on annotations.Rmd:
```diff
@@ -228,28 +249,113 @@ ggplot(economics, aes(date, unemploy)) +
+
+## Direct labelling
+
+The Subaru plots above provide examples of "direct labelling", in which the plot region itself contains the labels for groups of points instead of using a legend.  This usually makes the plot easier to read because it puts the labels closer to the data. The broader ggplot2 ecosystem contains a variety of other tools to accomplish this in a more automated fashion. The [directlabels](https://github ...
```
I think we need to decide on a style for referencing other packages. I think it's useful for web readers to include the link, but maybe we should replace "by Toby Dylan Hocking" with a formal reference?

djnavarro replied:
Agreed. I don't have any preference for how to approach this (and so have not done anything about it as yet) but it would be nice to do this in a consistent fashion throughout

djnavarro commented:
Agreed we could (& should) revisit this later to smooth: e.g. at the moment the chapter ends abruptly with no wrap up, and I haven't addressed the citation issue. That said, I share the intuition that it makes sense to revisit other chapters and then see how it all fits together. 

Assuming this looks okay to merge I'll likely browse the issues and pick out something mechanical to do next. Probably the cross-linking issue :-)

hadley commented:
Sounds great, thanks @djnavarro !
